### PR TITLE
Rework index map to add introspectibility and fix outstanding bug

### PR DESCRIPTION
**Pull request by collijk**
### Description

*Category*: bugfix
*JIRA issue*: 

**Main Changes**:
The index map used to keep a series whose index was the key columns and values were the index into the randomness system.  E.g.
```
crn_attr1   crn_attr2
2005-07-01  626880       650898
2005-07-01  383226       156931
2005-07-01  608111       893377
2005-07-01  993457       883587
2005-07-01  57456        800131
                                          ...  
2005-07-01  756277       946883
2005-07-01  871183       967051
2005-07-01  675501       395625
2005-07-01  377531       897957
2005-07-01  590229       915907
```
This has been updated to also keep a hold of the population index.  E.g.:
```
simulant_index  crn_attr1   crn_attr2
0               2005-07-01  626880       650898
1               2005-07-01  383226       156931
2               2005-07-01  608111       893377
3               2005-07-01  993457       883587
4               2005-07-01  57456        800131
                                          ...  
95              2005-07-01  756277       946883
96              2005-07-01  871183       967051
97              2005-07-01  675501       395625
98              2005-07-01  377531       897957
99              2005-07-01  590229       915907
```

This is done for two reasons:
1. It's much easier to introspect and compare fore debugging
2. It helps fix an outstanding bug in `IndexMap.__getitem__` (see https://github.com/ihmeuw/vivarium/pull/268 for more description) by allowing us to explicitly use `.loc` in the map indexer rather than relying on the inconsistent behavior of `pd.Series.__getitem__` depending on the underlying dtypes of the index.  The corresponding tests (`tests/framework/randomness/test_crn.py::test_failure_path_when_first_crn_attribute_not_datelike`) are no longer marked `xfail`.

**Other Changes**
1. The IndexMap has had it's methods and docstrings refactored for more clarity.
2. Randomness `key_columns` are now passed directly to the IndexMap and `use_crn` is inferred in the `__init__`
3. `RandomnessStream.get_draw` no longer passes the index map to the underlying `random` function when being used to initialize CRN variables.  This is a workaround for some bad error handling and anticipates some additional changes coming in the next PR to address these more directly.
4. The `metadata` function in `testing_utilities` has been expanded to take a layer argument (in a backwards compatible way as it's used in other libraries). 
5. The layer argument is then used in the main vivarium `conftest.py` to lower the priority of the fixture config so it can be overridden in other tests.
6. This override behavior is used to turn off crn in the state machine tests.  It wasn't being used before but was relying on bad error handling (see 3.) rather than properly declaring it didn't need CRN.
7. Various tests have been updated to reflect updates to the `IndexMap` public and private interfaces.

### Testing
CI and fixed regression test

### CODE REVIEW

File: tests/framework/randomness/test_crn.py [hunk 1]
```diff
@@ -280,8 +280,8 @@ def on_initialize_simulants(self, pop_data: SimulantData):
 @pytest.mark.parametrize(
     "with_crn, sims_to_add",
     [
-        pytest.param(True, cycle([0]), marks=pytest.mark.xfail),
-        pytest.param(True, cycle([1]), marks=pytest.mark.xfail),
+        pytest.param(True, cycle([0])),
+        pytest.param(True, cycle([1])),
```
Comment (rmudambi) [inline tests/framework/randomness/test_crn.py:284]: You should rename `BrokenPopulation`, since it is no longer broken.

**Review by rmudambi — APPROVED** [review]

File: src/vivarium/framework/randomness/index_map.py [hunk 2]
```diff
@@ -158,8 +243,8 @@ def _shift(self, m: Union[float, pd.Series]) -> Union[int, pd.Series]:
     def __getitem__(self, index: pd.Index) -> pd.Series:
         if not self._use_crn:
             return pd.Series(index, index=index)
-        if isinstance(index, (pd.Index, pd.MultiIndex)):
-            return self._map[index]
+        if isinstance(index, pd.Index):
```
[the remark this answers is not in the document]
Comment (collijk) [reply src/vivarium/framework/randomness/index_map.py:246]: yes

File: src/vivarium/framework/randomness/index_map.py [hunk 3]
```diff
@@ -21,42 +21,127 @@
...
+            new mapping index contains only the values for the new keys and the final mapping
+            combines the existing mapping and the new mapping index.
+
+        """
+        keys = new_keys.copy()
+        keys.index.name = self.SIM_INDEX_COLUMN
+        new_mapping_index = keys.set_index(self._key_columns, append=True).index
+
+        if self._map is not None:
+            final_mapping_index = self._map.index.append(new_mapping_index)
```
[the remark this answers is not in the document]
Comment (collijk) [reply src/vivarium/framework/randomness/index_map.py:82]: Duplicates are checked outside this function on line 49

File: tests/framework/randomness/test_crn.py [hunk 4]
```diff
@@ -255,8 +255,12 @@ def test_multi_sim_reproducibility_with_different_pop_growth(with_crn, pop_class
         assert_frame_equal(pop1.loc[overlap], pop2.loc[overlap])
 
 
-class BrokenPopulation(BasePopulation):
-    """CRN system falls over if the first CRN attribute is an int or float."""
+class UnBrokenPopulation(BasePopulation):
```
Comment (rmudambi) [inline tests/framework/randomness/test_crn.py:258]: Lol

**Review by rmudambi — APPROVED** [review]

**Pull request merged**